apps/emqx/src/emqx_packet.erl

```diff
@@ -640,20 +650,14 @@ format_payload_limit(Type0, Payload, Limit) when size(Payload) > Limit ->
             do_format_payload(Type, Payload)
     end;
 format_payload_limit(text, Payload, _Limit) ->
-    case is_utf8(Payload) of
-        true ->
-            do_format_payload(text, Payload);
-        false ->
-            do_format_payload(hex, Payload)
-    end;
+    do_format_payload(text, Payload);
 format_payload_limit(hex, Payload, _Limit) ->
     do_format_payload(hex, Payload).
 
 do_format_payload(text, Bytes) ->
-    %% utf8 ensured
-    Bytes;
+    unicode:characters_to_list(Bytes);
 do_format_payload(hex, Bytes) ->
-    ["hex:", binary:encode_hex(Bytes)].
+    binary_to_list(binary:encode_hex(Bytes)).
```

Comment:
Both this module's export mention `iolist()` is returned, and `emqx_trace_formatter` correctly handles that. Maybe it's `emqx_logger_textfmt` that [needs `unicode:characters_to_list/1` call](https://github.com/emqx/emqx/blob/74c346f9d1c063e529bbcf419096610ccdbd18ef/apps/emqx/src/emqx_logger_textfmt.erl#L192-L193) instead?

Reply:
we need the `hex` here because the function `format_payload/2` is called from other places. e.g. emqx_trace_json_formatter.erl

Reply:
To be clear, I initially commented on all these `*_to_list` conversions.

Reply:
Fixed.